src/blockchain_utilities/blockchain_prune.cpp

```diff
@@ -0,0 +1,697 @@
+  if (cmp)
+    ((flags & MDB_DUPSORT) ? mdb_set_dupsort : mdb_set_compare)(txn0, dbi0, cmp);
+
+  dbr = mdb_dbi_open(txn1, table, flags, &dbi1);
+  if (dbr) throw std::runtime_error("Failed to open LMDB dbi: " + std::string(mdb_strerror(dbr)));
+  if (cmp)
+    ((flags & MDB_DUPSORT) ? mdb_set_dupsort : mdb_set_compare)(txn1, dbi1, cmp);
+
+  dbr = mdb_txn_commit(txn1);
+  if (dbr) throw std::runtime_error("Failed to commit txn: " + std::string(mdb_strerror(dbr)));
+  tx_active1 = false;
+  MDB_stat stats;
+  dbr = mdb_env_stat(env0, &stats);
+  if (dbr) throw std::runtime_error("Failed to stat " + std::string(table) + " LMDB table: " + std::string(mdb_strerror(dbr)));
+  check_resize(env1, (stats.ms_branch_pages + stats.ms_overflow_pages + stats.ms_leaf_pages) * stats.ms_psize);
+  dbr = mdb_txn_begin(env1, NULL, 0, &txn1);
+  if (dbr) throw std::runtime_error("Failed to create LMDB transaction: " + std::string(mdb_strerror(dbr)));
+  tx_active1 = true;
+
+  dbr = mdb_drop(txn1, dbi1, 0);
+  if (dbr) throw std::runtime_error("Failed to empty " + std::string(table) + " LMDB table: " + std::string(mdb_strerror(dbr)));
+
+  dbr = mdb_cursor_open(txn0, dbi0, &cur0);
+  if (dbr) throw std::runtime_error("Failed to create LMDB cursor: " + std::string(mdb_strerror(dbr)));
+  dbr = mdb_cursor_open(txn1, dbi1, &cur1);
+  if (dbr) throw std::runtime_error("Failed to create LMDB cursor: " + std::string(mdb_strerror(dbr)));
+
+  MDB_val k;
+  MDB_val v;
+  MDB_cursor_op op = MDB_FIRST;
+  size_t nrecords = 0, bytes = 0;
+  while (1)
+  {
+    int ret = mdb_cursor_get(cur0, &k, &v, op);
+    op = MDB_NEXT;
+    if (ret == MDB_NOTFOUND)
+      break;
+    if (ret)
+      throw std::runtime_error("Failed to enumerate " + std::string(table) + " records: " + std::string(mdb_strerror(ret)));
+
+    bytes += k.mv_size + v.mv_size;
+    if (resize_point(++nrecords, env1, &txn1, bytes))
+    {
+      dbr = mdb_cursor_open(txn1, dbi1, &cur1);
+      if (dbr) throw std::runtime_error("Failed to create LMDB cursor: " + std::string(mdb_strerror(dbr)));
+    }
+
+    ret = mdb_cursor_put(cur1, &k, &v, putflags);
+    if (ret)
+      throw std::runtime_error("Failed to write " + std::string(table) + " record: " + std::string(mdb_strerror(ret)));
+  }
+
+  mdb_cursor_close(cur1);
+  mdb_cursor_close(cur0);
+  mdb_txn_commit(txn1);
+  tx_active1 = false;
+  mdb_txn_commit(txn0);
+  tx_active0 = false;
+  mdb_dbi_close(env1, dbi1);
+  mdb_dbi_close(env0, dbi0);
+}
+
+static bool is_v1_tx(MDB_cursor *c_txs_pruned, MDB_val *tx_id)
+{
+  MDB_val v;
+  int ret = mdb_cursor_get(c_txs_pruned, tx_id, &v, MDB_SET);
+  if (ret)
+    throw std::runtime_error("Failed to find transaction pruned data: " + std::string(mdb_strerror(ret)));
+  if (v.mv_size == 0)
+    throw std::runtime_error("Invalid transaction pruned data");
+  uint64_t version;
+  std::string tmp((const char*)v.mv_data, v.mv_size);
```

Comment:
And removing the second const doens't help.

Reply:
They both have to be l-values due to the parameter list. And the second const would break the instantiation stage:
```c++
const char* begin = static_cast<const char*>(v.mv_data);
const char* end = begin + v.mv_size;
read_varint(begin, end, version);
```

Reply:
This version doesn't compile either.

<pre>
  bool is_v1_tx(const blobdata_ref& tx_blob)
  {
    uint64_t version;
    const char* begin = static_cast<const char*>(tx_blob.data());
    const char* end = begin + tx_blob.size();
    int read = tools::read_varint(begin, end, version);
    if (read <= 0)
      throw std::runtime_error("Internal error getting transaction version");
    return version <= 1;
  }
</pre>

<pre>
[ 31%] Building CXX object src/cryptonote_basic/CMakeFiles/obj_cryptonote_basic.dir/cryptonote_format_utils.cpp.o
In file included from /home/user/src/bitmonero/src/cryptonote_basic/cryptonote_format_utils.cpp:36:
In file included from /home/user/src/bitmonero/src/cryptonote_basic/cryptonote_format_utils.h:33:
In file included from /home/user/src/bitmonero/src/cryptonote_basic/cryptonote_basic_impl.h:33:
In file included from /home/user/src/bitmonero/src/cryptonote_basic/cryptonote_basic.h:41:
In file included from /home/user/src/bitmonero/src/serialization/binary_archive.h:41:
/home/user/src/bitmonero/src/common/varint.h:126:12: error: no matching function for call to 'read_varint'
    return read_varint<std::numeric_limits<T>::digits, InputIt, T>(std::move(first), std::move(last), i);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/user/src/bitmonero/src/cryptonote_basic/cryptonote_format_utils.cpp:241:23: note: in instantiation of function template specialization 'tools::read_varint<const char *&, unsigned long>' requested here
    int read = tools::read_varint(begin, end, version);
                      ^
/home/user/src/bitmonero/src/common/varint.h:94:5: note: candidate function not viable: no known conversion from 'typename std::remove_reference<const char *&>::type' (aka 'const char *') to 'const char *&&' for 1st argument
    read_varint(InputIt &&first, InputIt &&last, T &write) {
    ^
/home/user/src/bitmonero/src/common/varint.h:125:9: note: candidate template ignored: invalid explicitly-specified argument for template parameter 'InputIt'
    int read_varint(InputIt &&first, InputIt &&last, T &i) {
        ^
1 error generated.
</pre>

Reply:
blobdata_ref is a epee::span<const char>

Reply:
Its a bug in the overload (2nd version):

```c++
return read_varint<std::numeric_limits<T>::digits>(std::forward<InputIt>(first), std::forward<InputIt>(last), i);
```
That one did compile, at least with clang 6.

Reply:
Thanks, that worked eventually. CLANG 6 too though.